### PR TITLE
adding lerna

### LINKED ISSUE #20: private settings in package.json
For [publishing to npm](https://docs.npmjs.com/cli/v7/configuring-npm/package-json#private), private must be set to false.

But [workspaces to work properly](https://blog.kintone.io/entry/2020/07/14/103322), private must be set to true.

[Lerna](https://www.npmjs.com/package/lerna) helps solve this problem for monorepos.  

@Dhaiwat10 please assign this to me

### PR DESCRIPTION
**Pull request by e-roy**
This will add lerna for easier publishing to npm.  This creates a massive change in the file structure of the project

### CODE REVIEW

**Comment by e-roy** [comment]
Should be linked to this issue:

[https://github.com/Developer-DAO/create-dao/issues/20](https://github.com/Developer-DAO/create-dao/issues/20)

**Comment by e-roy** [comment]
@with-heart I just noticed the .bin files in my PR.  Should I delete them?

**Comment by Dhaiwat10** [comment]
> @with-heart I just noticed the .bin files in my PR. Should I delete them?

I think you should

File: package.json
Comment (Dhaiwat10) [inline package.json]: What does this command do? Does it execute the script that the users will execute? i.e. Can this command be used to test out any changes locally?

**Review by Dhaiwat10 — APPROVED** [review]
LGTM :) Just left one little question.

File: package.json
Comment (e-roy) [reply package.json]: The "deploy" script is to deploy contracts for the create-dao workspace
The "update-version" script will update the npm packages for publishing
The  "publish-latest" script will publish the packages
The "start:npx:create-dao" script (I just changed the name in latest update)  will run a test of create-dao locally

I just added docs to this PR, please let me know if the docs are clear, doc writing isn't a strength :)

**Comment by with-heart** [comment]
I promise I will review this today! I'm sorry it is taking so long

File: docs/publish-npm.md
Comment (revmischa) [inline docs/publish-npm.md]:
```suggestion
- Important - when adding a new package, be sure to remove the private attribute from the package.json file of that package.
```
Right? "private key" sounds confusing

Comment (e-roy) [reply docs/publish-npm.md]: agree, thank you for looking over

File: docs/local-development.md
Comment (with-heart) [inline docs/local-development.md:10]: It'd be great if we could consolidate this into a single `start` script if possible. Could be a nice improvement to the DX! 

Will open a follow-up issue to investigate that.

File: package.json
Comment (with-heart) [inline package.json]:
```suggestion
    "create-dao": "node ./packages/create-dao/index.js"
```

imo it makes sense to just call this `create-dao`, as if you're running the actual binary. It doesn't really have anything to do with `start` or `npx`, ya know?

File: package.json
Comment (with-heart) [inline package.json]: I think it's fairly typical for the root `package.json` to have a `name` that matches the repo name. Not a big deal but just wanted to throw that out there!

File: package.json
Comment (with-heart) [inline package.json]:
```suggestion
    "publish-latest": "lerna publish from-package",
```

File: lerna.json
Comment (with-heart) [inline lerna.json]:
```suggestion
  "packages": ["packages/*"],
```

We need this to be a glob for the "Adding more workspaces" instructions to work currently (since there's no instruction to add a new workspace to this array).

It also means those instructions can be more specific by telling contributors to create a new workspace specifically in the `packages` folder.

File: docs/local-development.md
Comment (with-heart) [inline docs/local-development.md:20]: If we update this instruction to tell the user to create the new workspace in the `packages` folder, they won't have to touch the root `package.json` at all since you're using the `packages/*` glob. I think that'd be preferable.

File: docs/publish-npm.md
Comment (with-heart) [inline docs/publish-npm.md]: I'm not really sure any of this belongs in this file. None of it is relevant to versioning or publishing. Plus the 2nd sentence doesn't even match the instructions in `local-development.md`, so I think this could be confusing.

**Review by with-heart — CHANGES_REQUESTED** [review]
We can proceed with it but I'm not sure how I feel about creating a separate `docs` folder!

The less places contributors have to look to find information, the more likely they are to find it. There's nothing wrong with having a fairly long `README.md`, especially this early in the project's life since we haven't really talked about documentation patterns yet.

It's also not really a required change to implement `lerna`. What I typically recommend is that projects try to keep PRs as "atomic" as possible, where each PR is essentially implementing/reviewing/discussing a single topic. 

Like let's say we decided to reject the `docs` folder idea. If that was a separate PR, rejecting is as simple as closing that PR. If it's just a piece of a larger PR (like this one), rejecting would mean more work for you to remove it and for us to re-review and more time overall for the PR to be merged.

Not meaning to pick on the `docs` folder in particular because it's really not a big deal and there's no problem in moving forward with it! Just trying to impart some context on how I view things as a maintainer :)

---

Just a few comments/issues, looks good but at least one or two of them should be addressed before we can merge!

File: package.json
Comment (e-roy) [reply package.json]: The package.json file in the create-dao package has the "create-dao" name already.

File: package.json
Comment (e-roy) [reply package.json]: good catch

File: docs/local-development.md
Comment (e-roy) [reply docs/local-development.md:10]: I think it would be great to consolidate into a single script, I didn't try this for this PR because it was outside the scope of it.  Please do open a issue for it

File: docs/local-development.md
Comment (e-roy) [reply docs/local-development.md:20]: The glob you are referencing is inside the lerna.json file which is for publishing to npm.  For creating a workspace for development, this will help to identify separate templates inside a package

File: docs/publish-npm.md
Comment (e-roy) [reply docs/publish-npm.md]: This is to be able to test a simulation of npx create-dao locally before publishing to find errors.  This has nothing to do with local development of the packages and/or templates themselves.

**Comment by e-roy** [comment]
I will be happy to remove the docs folder.  I will also be happy to use another solution than implementing lerna if desired.  
 
It seemed this project was inspired by create-next-app as from the discussion  [here](https://github.com/Developer-DAO/create-dao/issues/3)
 
I noticed by going with a wrapper design it created this issue [here](https://github.com/Developer-DAO/create-dao/issues/12)
 
So I started to reverse engineer how CNA did it and arrived to this solution [here](https://github.com/Developer-DAO/create-dao/pull/17)
 
Which created this issue [here](https://github.com/Developer-DAO/create-dao/issues/18)
 
To solve it with workspaces [here](https://github.com/Developer-DAO/create-dao/pull/19)
 
Which this lead to this issue [here](https://github.com/Developer-DAO/create-dao/issues/20)
 
To resolve with the solution in this current PR
 
I created the docs folder because of discussions from our discord for having better documentation, also from @Dhaiwat10 comment in this thread that there was a need to elaborate further on scripts.  
Inside the README.md there are links that referenced these doc files for further context for contributors to find everything in one organized place. 
 
All of which are inspired from CNA's design, making the assumption from their product's success, was something worth modeling from to create a scalable product.
 
I also assumed that since this product was still far from an MVP that large rapid changes in the beginning would get to product market fit faster for further development, and I apologize for that assumption also.
 
I will happily implement smaller changes with more discussions if that is what is desired.
 
Please advise on whether I should continue with lerna, or implement another design for this project, and also the removal of the docs folder.  I’m looking forward to helping and contributing to the community :)

**Comment by Dhaiwat10** [comment]
@e-roy I think we're good!

**Comment by with-heart** [comment]
Ye, definitely good to go!

Had been working on a reply but hadn't finished it yet. Wanted to go ahead and post it anyway :)

---

I don't think you really have anything to apologize for! Like I said before, we can definitely move forward with what you have, just wanted to share some context about how I tend to think about things. 

Your reasoning makes perfect sense and I appreciate you taking the time to provide the past context that lead you to this PR! ❤️ 

> All of which are inspired from CNA's design, making the assumption from their product's success, was something worth modeling from to create a scalable product.

For sure! I think that's a good point and is a logical assumption to make. I think it's something worth modeling _as we mature and scale as a product_, but as it is right now, this is a much smaller project with a much less experienced team building it, so it's good to build our processes with that in mind too! 

The `docs` folder really isn't as big of a deal as some of my comments may have made it out to be and I'm sorry for that! My thinking was that with such a small team and lots of potentially newbie contributors, the less we spread information out the easier it is to find that information. But a few extra files is not a problem, especially since you linked them from `README.md`!

> I also assumed that since this product was still far from an MVP that large rapid changes in the beginning would get to product market fit faster for further development, and I apologize for that assumption also.
>
> I will happily implement smaller changes with more discussions if that is what is desired.

I think that's a good assumption to make and I'm definitely in favor of focusing on action and moving fast over spending too much time on process/discussion. 

At the same time, balance is usually a good thing and I think the more we can break things down, the easier it is to have more targeted discussions, more flexible outcomes, and better historical documentation of what we're doing, why we're doing it, and how we did it!

imo it's less of a hard requirement in my view and more like "where convenient, prefer to break things down". Sometimes it's convenient, sometimes it isn't. We'll work through it either way :)

**Pull request merged**
